Recursive extract_all_objects() not working as expected on macOS

I tried using extract_all_objects()  instead of a helper library + link_whole for a test, but I got a lot of linker errors with that:

```
[57/62] Linking target tests/checkasm.
FAILED: tests/checkasm 
cc  -o tests/checkasm 'tests/tests@@checkasm@exe/checkasm_checkasm.c.o' 'tests/tests@@checkasm@exe/checkasm_mc.c.o' 'dav1d_bitdepth_8@sta/src_ipred.c.o' 'dav1d_bitdepth_8@sta/src_itx.c.o' 'dav1d_bitdepth_8@sta/src_ipred_prepare.c.o' 'dav1d_bitdepth_8@sta/src_lf_apply.c.o' 'dav1d_bitdepth_8@sta/src_loopfilter.c.o' 'dav1d_bitdepth_8@sta/src_mc.c.o' 'dav1d_bitdepth_8@sta/src_cdef_apply.c.o' 'dav1d_bitdepth_8@sta/src_cdef.c.o' 'dav1d_bitdepth_8@sta/src_lr_apply.c.o' 'dav1d_bitdepth_8@sta/src_looprestoration.c.o' 'dav1d_bitdepth_8@sta/src_recon.c.o' 'dav1d_bitdepth_8@sta/src_x86_mc_init.c.o' 'dav1d_bitdepth_10@sta/src_ipred.c.o' 'dav1d_bitdepth_10@sta/src_itx.c.o' 'dav1d_bitdepth_10@sta/src_ipred_prepare.c.o' 'dav1d_bitdepth_10@sta/src_lf_apply.c.o' 'dav1d_bitdepth_10@sta/src_loopfilter.c.o' 'dav1d_bitdepth_10@sta/src_mc.c.o' 'dav1d_bitdepth_10@sta/src_cdef_apply.c.o' 'dav1d_bitdepth_10@sta/src_cdef.c.o' 'dav1d_bitdepth_10@sta/src_lr_apply.c.o' 'dav1d_bitdepth_10@sta/src_looprestoration.c.o' 'dav1d_bitdepth_10@sta/src_recon.c.o' 'dav1d_bitdepth_10@sta/src_x86_mc_init.c.o' 'libdav1dentrypoint@sta/src_lib.c.o' 'libdav1dentrypoint@sta/src_thread_task.c.o' 'dav1d_private@sta/src_picture.c.o' 'dav1d_private@sta/src_data.c.o' 'dav1d_private@sta/src_ref.c.o' 'dav1d_private@sta/src_getbits.c.o' 'dav1d_private@sta/src_obu.c.o' 'dav1d_private@sta/src_decode.c.o' 'dav1d_private@sta/src_cdf.c.o' 'dav1d_private@sta/src_msac.c.o' 'dav1d_private@sta/src_tables.c.o' 'dav1d_private@sta/src_scan.c.o' 'dav1d_private@sta/src_dequant_tables.c.o' 'dav1d_private@sta/src_intra_edge.c.o' 'dav1d_private@sta/src_lf_mask.c.o' 'dav1d_private@sta/src_ref_mvs.c.o' 'dav1d_private@sta/src_warpmv.c.o' 'dav1d_private@sta/src_wedge.c.o' 'dav1d_private@sta/src_qm.c.o' 'dav1d_private@sta/src_x86_cpu.c.o' -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-headerpad_max_install_names  
Undefined symbols for architecture x86_64:
  "_dav1d_avg_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_cpu_cpuid", referenced from:
      _get_cpu_flags in src_x86_cpu.c.o
  "_dav1d_cpu_xgetbv", referenced from:
      _get_cpu_flags in src_x86_cpu.c.o
  "_dav1d_mask_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_regular_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_regular_sharp_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_regular_smooth_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_sharp_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_sharp_regular_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_sharp_smooth_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_smooth_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_smooth_regular_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_8tap_smooth_sharp_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_prep_bilin_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_regular_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_regular_sharp_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_regular_smooth_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_sharp_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_sharp_regular_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_sharp_smooth_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_smooth_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_smooth_regular_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_8tap_smooth_sharp_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_put_bilin_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_w_avg_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
  "_dav1d_w_mask_420_avx2", referenced from:
      _dav1d_mc_dsp_init_x86_8bpc in src_x86_mc_init.c.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[58/62] Generating 'dav1d_private@sta/mc.obj'.
ninja: build stopped: subcommand failed.
```

The code can be found here: https://code.videolan.org/ePirat/dav1d/tree/gramner-twoorioles/dav1d-checkasm

--- a/mesonbuild/backend/backends.py
+++ b/mesonbuild/backend/backends.py
@@ -441,7 +441,13 @@ class Backend:
                 sources.append(File(True, dirpart, fnamepart))
 
         # Filter out headers and all non-source files
-        sources = [s for s in sources if self.environment.is_source(s) and not self.environment.is_header(s)]
+        filtered_sources = []
+        for s in sources:
+            if self.environment.is_source(s) and not self.environment.is_header(s):
+                filtered_sources.append(s)
+            elif self.environment.is_object(s):
+                result.append(s.relative_name())
+        sources = filtered_sources
 
         # extobj could contain only objects and no sources
         if not sources:

--- a/test cases/nasm/1 configure file/meson.build
+++ b/test cases/nasm/1 configure file/meson.build
@@ -47,3 +47,9 @@ exe = executable('hello', asm_gen.process('hello.asm'),
 )
 
 test('test-nasm-configure-file', exe)
+
+exe2 = executable('hello2', objects : exe.extract_all_objects(),
+  link_args: link_args,
+)
+
+test('test-nasm-extract-all-objects', exe2)

